ext/yajl/yajl_ext.c

```diff
@@ -560,6 +566,385 @@ static VALUE rb_yajl_parser_set_complete_cb(VALUE self, VALUE callback) {
+            val = rb_yajl_projector_filter_subtree(parser, schema, event);
+        } else {
+            val = rb_yajl_projector_build_simple_value(parser, event);
+        }
+        rb_ary_push(ary, val);
+
+        event = yajl_event_stream_next(parser, 0);
+        if (event.token == yajl_tok_comma) {
+            event = yajl_event_stream_next(parser, 1);
+            assert(event.token == yajl_tok_comma);
+
+            event = yajl_event_stream_next(parser, 0);
+            if (!(event.token == yajl_tok_string || event.token == yajl_tok_integer || event.token == yajl_tok_double || event.token == yajl_tok_null || event.token == yajl_tok_bool || event.token == yajl_tok_left_bracket || event.token == yajl_tok_left_brace)) {
+                rb_raise(cParseError, "read a comma, expected a value to follow, actually read %s", yajl_tok_name(event.token));
+            }
+        } else if (event.token != yajl_tok_right_brace) {
+            rb_raise(cParseError, "didn't read a comma, expected closing array, actually read %s", yajl_tok_name(event.token));
+        }
+    }
+
+    return ary;
+}
+
+static VALUE rb_yajl_projector_filter_object_subtree(yajl_event_stream_t parser, VALUE schema, yajl_event_t event) {
+    assert(event.token == yajl_tok_left_bracket);
+
+    VALUE hsh = rb_hash_new();
+
+    while (1) {
+        event = yajl_event_stream_next(parser, 1);
+
+        if (event.token == yajl_tok_right_bracket) {
+            break;
+        }
+
+        if (!(event.token == yajl_tok_string || event.token == yajl_tok_string_with_escapes)) {
+            rb_raise(cParseError, "Expected string, unexpected stream event %s", yajl_tok_name(event.token));
+        }
+
+        VALUE key = rb_yajl_projector_build_string(parser, event);
+
+        event = yajl_event_stream_next(parser, 1);
+        if (!(event.token == yajl_tok_colon)) {
+            rb_raise(cParseError, "Expected colon, unexpected stream event %s", yajl_tok_name(event.token));
+        }
+
+        // nil schema means reify the subtree from here on
+        // otherwise if the schema has a key for this we want it
+        int interesting = (schema == Qnil || rb_funcall(schema, rb_intern("key?"), 1, key) == Qtrue);
+        if (!interesting) {
+            rb_yajl_projector_ignore_value(parser);
+            goto peek_comma;
+        }
+
+        yajl_event_t value_event = yajl_event_stream_next(parser, 1);
+
+        VALUE val;
+        if (value_event.token == yajl_tok_left_bracket || value_event.token == yajl_tok_left_brace) {
+            VALUE key_schema;
+            if (schema == Qnil) {
+                key_schema = Qnil;
+            } else {
+                key_schema = rb_hash_aref(schema, key);
+            }
+
+            val = rb_yajl_projector_filter_subtree(parser, key_schema, value_event);
+        } else {
+            val = rb_yajl_projector_build_simple_value(parser, value_event);
+        }
+        
+        rb_hash_aset(hsh, key, val);
```

Comment:
I think we might want to call `rb_str_freeze` on the key here.  Otherwise `rb_hash_aset` will [dup and freeze the key](https://github.com/ruby/ruby/blob/5fd6558502b30e1450023b73a7bbbd8f419767c7/hash.c#L1572-L1582).